geopb: make SpatialObject.MemSize a bit more precise

yuzefovich opened the pull request:
Previously, we only included the length of the serialized byte slice, but now we'll include the capacity (which is a better representation of the memory footprint). For example, this will provide better initial estimate in `st_memcollect` aggregate builtins. In some experimental testing this estimate appears to be an over-estimate (instead of an under-estimate as it used to be) which seems beneficial (more stability).

There is a caveat, though, that we expose the size via `st_memsize` builtin, and in order to make it return deterministic results (i.e. independent of how the memory is allocated), we'll preserve the "deterministic size" function too (only used for `st_memsize` and in tests).

Fixes: #124232.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/124594)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124594#-:-Nyco10Z6bgRSsjHxyRD:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
geopb: make SpatialObject.MemSize a bit more precise

Previously, we only included the length of the serialized byte slice,
but now we'll include the capacity (which is a better representation of
the memory footprint). For example, this will provide better initial
estimate in `st_memcollect` aggregate builtins. In some experimental
testing this estimate appears to be an over-estimate (instead of an
under-estimate as it used to be) which seems beneficial (more
stability).

There is a caveat, though, that we expose the size via `st_memsize`
builtin, and in order to make it return deterministic results (i.e.
independent of how the memory is allocated), we'll preserve the
"deterministic size" function too (only used for `st_memsize` and in
tests).

Release note: None

yuzefovich reviewed (commented):
I had to tweak the PR slightly to make `st_memsize` deterministic - would appreciate another quick look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124594#-:-Nz1WF3B4Hb3byVbhYjg:bif870l)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Good catch. Makes sense!

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124594#-:-Nz1yioR0mbinR1vmy41:b6n8jfr)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
